src/entry.c

```diff
@@ -0,0 +1,364 @@
+    bool create_new_entry = (update_expiry && (entryGetExpiry(e) == EXPIRY_NONE || ttl == EXPIRY_NONE)) ||
+                            !(update_value && !entryHasValuePtr(e) &&
+                              required_embedded_size <= EMBED_VALUE_MAX_ALLOC_SIZE &&
+                              required_embedded_size <= current_embedded_allocation_size &&
+                              required_embedded_size >= current_embedded_allocation_size * 3 / 4);
+
+    if (!create_new_entry) {
+        /* In this case we are sure we do not have to allocate new entry, so expiry must already be set. */
+        if (update_expiry) {
+            serverAssert(entryHasExpiry(e));
+            char *buf = sdsAllocPtr(e);
+            if (entryHasValuePtr(e)) buf -= sizeof(sds *);
+            buf -= sizeof(expiry);
+            memcpy(buf, &expiry, sizeof(expiry));
+        }
+        /* In this case we are sure we do not have to allocate new entry, so value must already be set or we have enough room to embed it. */
+        if (update_value) {
+            if (entryHasValuePtr(e)) {
+                sds *value_ref = entryGetValueRef(e);
+                sdsfree(*value_ref);
+                *value_ref = value;
+            } else {
+                /* Skip field content, field null terminator and value sds8 hdr. */
+                sds old_value = entryGetValue(e);
+                sdswrite(sdsAllocPtr(old_value), sdsAllocSize(old_value), SDS_TYPE_8, value, sdslen(value));
+                sdsfree(value);
+            }
+        }
+        return e;
+
+    } else {
+        if (!update_value) {
+            /* Check if the value can be reused. */
+            int value_was_embedded = !entryHasValuePtr(e);
+            /* In case the original entry value is embedded WE WILL HAVE TO DUPLICATE IT */
+            if (value_was_embedded)
+                value = sdsdup(value);
+            /* if not we have to duplicate it, remove it from the original entry since we are going to delete it.*/
+            else {
+                sds *value_ref = entryGetValueRef(e);
+                *value_ref = NULL;
+            }
+        }
+    }
+
+    entry *new_entry = entryCreate(entryGetField(e), value, ttl);
+    if (new_entry != e)
+        entryFree(e);
+    return new_entry;
+}
+
+/* Returns memory usage of a entry, including all allocations owned by
+ * the entry. */
+size_t entryMemUsage(entry *entry) {
+    size_t mem = 0;
+
+    if (entryHasValuePtr(entry)) {
+        /* In case the value is not embedded we might not be able to sum all the allocation sizes since the field
+         * header could be too small for holding the real allocation size. */
+        mem += zmalloc_usable_size(entryAllocPtr(entry));
+    } else {
+        mem += sdsReqSize(sdslen(entry), sdsType(entry));
+        if (entryHasExpiry(entry)) mem += sizeof(long long);
+    }
+    mem += sdsAllocSize(entryGetValue(entry));
+    return mem;
```

Comment:
This doesn't really make sense to me. I think there need to be unit tests.
1. I think line 330 is wrong for embedded values because sdsAllocSize doesn't work on embedded sds strings - it uses malloc functions that assume the sds has a dedicated allocation.
2. Suppose we have an embedded value entry, `updateEntry()` was called with a slightly smaller value but it kept the allocation because `required_embedded_size >= current_embedded_allocation_size * 3 / 4`. You have to use something like zmalloc_usable_size() because the size of the members no longer adds up to the size of the allocation they're stored in.

```suggestion
    size_t mem = zmalloc_usable_size(entryAllocPtr(entry));

    if (entryHasValuePtr(entry)) {
        /* Value is separate allocation when not embedded */
        mem += sdsAllocSize(entryGetValue(entry));
    }
    return mem;
```

Reply:
this logic was not really changed from the previous implementation.
 in case the value is **NOT** embedded, we can go to the start of the allocation (either with TTL or not) and check the size of the allocation, and later add the value allocation size. in case the value **IS** embedded we calculate the elements sizes including the value. 
 The embedded value is always being assigned with the remaining allocation size in the buffer in entryCreate, so when we are NOT reallocating the entry this value should still be accurate right?

Reply:
I better understand `createEntry` and `updateEntry` now, so yes this makes sense and looks correct to me. I also understand now that values are supposed to never use `SDS_TYPE_5` so `sdsAllocSize(entryGetValue(entry))` is supposed to not invoke undefined behavior, but this wasn't obvious and won't be obvious to future contributors - I'd like to see a comment to clarify that tricky business here or we just avoid using the trick.

Curious why we avoid calling `zmalloc_usable_size` for the entry - is this code faster or some other benefit I don't see?

Reply:
zmalloc_usable_size will require some expensive arena metadata access which is better be avoided.

Reply:
added comment